content/blog/2019-12-13-futil-fsm/index.md

```diff
@@ -0,0 +1,315 @@
+| Control           | FSM             |
+
+### *Read* Signals
+
+In FuTIL semantics, `enable` keyword is used to determine whether a component is active. It is the easiest way of translating a program into hardware. However, this implicitly assumes that the signal on a wire is not valid or readable until we `enable` a component. We therefore require any *data* wire to have one extra bit to specify whether the signal is readable.
+
+### *MUX*
+
+A component can be used more than once in FuTIL. For instance, if we reuse variable `a0`, then the register need to choose the input from const0 and const2. This introduces a multiplexer (MUX).
+
+```futil
+(enable a0 const0)
+(enable a0 const2)
+```
+
+At different time step, read signals tells which wire to the MUX is readable. Therefore, read wires serve as **sel** signals for MUX.
+
+### *FSM*
+
+In FuTIL, there are control logics like `if`, `while` and etc. This can be translated into FSM in Verilog implementation, which is the main goal of this project. However, before getting to that, we created intermediate FSM expressions in FuTIL. An FSM component has:
+
+- input and output ports,
+- connection of wires between its own ports and other components's port,
+- internal control logic that determine the output signals.
+
+The internal control logic of a FSM component can be divided into several states that determines the output signals. A state transfers to another according to some input signals. In general, all FSM components are composed of one **Start** state, some **Intermediate** states and **End** state.
+
+Consider the syntax `(enable A B)` . The **Start** state transfers to the **Intermediate** when the *valid* signal is high. At the **Intermediate** state, the FSM sends out valid signals to subcomponents `A` and `B`, and waits for *ready* signals from them to be high. Once both of the *ready* signals are high, the FSM transfers to **End** state and outputs *ready* signals to notify upper components. It transfers back to **Start** state when *valid* signal is low, indicating the upper components have received the *ready* signal and finished execution so it is safe for the FSM to go back to the **Start** state. The same design logic applies to all FSMs. The only difference happens in intermediate state(s): `seq` FSM has one or more intermediate states and one intermediate only transfers to next state when receiving high *ready* signal from the previous state; `if` FSM send *valid* to the module that execute the comparison and receive both *ready* and *condition* signals and determine which state it should transfer to with the *condition* signal; `while` FSM transfers to loop **Body** state when *condition* signal is high and goes to **End** State when condition is low.
+
+<img src="fsm.png" style="width: 100%">
+
+
+
+
+
+## Implementation
+
+<img src="flow.png" style="width: 100%">
+
+At implementation, to realize what we describe in design overview, we gradually added intermediate passes.
```

Comment:
- maybe just delete "At implementation," which seems redundant?
- in design overview -> in the design overview